packages/styled-components/transforms/src/visitors/transpile_css_prop/top_level_binding_collector.rs

```diff
@@ -0,0 +1,100 @@
+use swc_atoms::JsWord;
+use swc_common::collections::AHashSet;
+use swc_ecmascript::{
+    ast::{
+        ArrowExpr, ClassDecl, FnDecl, Function, ImportDefaultSpecifier, ImportNamedSpecifier,
+        ImportStarAsSpecifier, ModuleItem, ObjectPatProp, Param, Pat, Stmt, VarDeclarator,
+    },
+    visit::{noop_visit_type, Visit, VisitWith},
+};
+
+// Modified from swc_ecma_utils/src/lib.rs:BindingCollector.
+pub struct TopLevelBindingCollector {
+    bindings: AHashSet<JsWord>,
+}
+
+impl TopLevelBindingCollector {
+    fn add(&mut self, i: &JsWord) {
+        self.bindings.insert(i.clone());
+    }
+}
+
+impl Visit for TopLevelBindingCollector {
+    noop_visit_type!();
+
+    fn visit_class_decl(&mut self, node: &ClassDecl) {
+        self.add(&node.ident.sym);
+    }
+
+    fn visit_fn_decl(&mut self, node: &FnDecl) {
+        self.add(&node.ident.sym);
+    }
+
+    fn visit_pat(&mut self, node: &Pat) {
+        match node {
+            Pat::Ident(i) => self.add(&i.id.sym),
+            Pat::Object(o) => {
+                for prop in o.props.iter() {
+                    match prop {
+                        ObjectPatProp::Assign(a) => self.add(&a.key.sym),
+                        ObjectPatProp::KeyValue(k) => k.value.visit_with(self),
+                        ObjectPatProp::Rest(_) => {}
+                    }
+                }
+            }
+            Pat::Array(a) => {
+                for elem in a.elems.iter() {
+                    elem.visit_with(self);
+                }
+            }
+            _ => {}
+        }
+    }
+
+    fn visit_param(&mut self, node: &Param) {
+        node.visit_children_with(self);
+    }
+
+    fn visit_arrow_expr(&mut self, _: &ArrowExpr) {}
+    fn visit_function(&mut self, _: &Function) {}
+
+    fn visit_import_default_specifier(&mut self, node: &ImportDefaultSpecifier) {
+        self.add(&node.local.sym);
+    }
+
+    fn visit_import_named_specifier(&mut self, node: &ImportNamedSpecifier) {
+        self.add(&node.local.sym);
+    }
+
+    fn visit_import_star_as_specifier(&mut self, node: &ImportStarAsSpecifier) {
+        self.add(&node.local.sym);
+    }
+
+    fn visit_module_items(&mut self, nodes: &[ModuleItem]) {
+        for node in nodes {
+            node.visit_children_with(self)
+        }
+    }
+
+    fn visit_stmts(&mut self, nodes: &[Stmt]) {
+        for node in nodes {
+            node.visit_children_with(self)
+        }
+    }
+
+    fn visit_var_declarator(&mut self, node: &VarDeclarator) {
+        node.name.visit_with(self);
+        node.init.visit_with(self);
+    }
+}
+
+pub fn collect_top_level_decls<N>(n: &N) -> AHashSet<JsWord>
```

Comment:
Side note, this will collect `a` in `a = foo;` (assignment).
Is this intended?

Reply:
Alternatively, you can use `swc_ecma_utils::collect_decls_with_ctxt(top_level_ctxt)`.
This will collect all top level bindings, while handling all concerns I mentioned.

Reply:
Also, binding collector should be using `Id`, not `JsWord`.

You can read more about variable management of swc at https://docs.rs/swc_ecma_ast/0.59.0/swc_ecma_ast/struct.Ident.html and https://docs.rs/swc_ecma_transforms_base/latest/swc_ecma_transforms_base/resolver/fn.resolver_with_mark.html

Reply:
Are the `SyntaxContext`s identical when an `Ident` declared in an outer scope is used from an inner scope? I think I made the `Ident` -> `JsWord` change because the `Ident`s didn't match in the `GlobalInterpolation` case.

Reply:
Yes. If two identifiers point same variable, `SyntaxContext` will be identical to each other.

Reply:
Looks like AHashSet lookup fails when looking up:

```
Ident {
    span: Span {
        lo: BytePos(
            3968,
        ),
        hi: BytePos(
            3974,
        ),
        ctxt: #1,
    },
    sym: Atom('radius' type=inline),
    optional: false,
}
```

in an `AHashSet` containing:

```
Ident {
    span: Span {
        lo: BytePos(
            3882,
        ),
        hi: BytePos(
            3888,
        ),
        ctxt: #1,
    },
    sym: Atom('radius' type=inline),
    optional: false,
},
```